python/pyarrow/_flight.pyx

```diff
@@ -135,18 +144,73 @@ cdef class FlightDescriptor:
     def __repr__(self):
         return "<FlightDescriptor type: {!r}>".format(self.descriptor_type)
 
+    @staticmethod
+    cdef FlightDescriptor unwrap(descriptor):
+        if not isinstance(descriptor, FlightDescriptor):
+            raise TypeError("Must provide a FlightDescriptor, not '{}'".format(
+                type(descriptor)))
+        return <FlightDescriptor> descriptor
+
 
-class Ticket:
+cdef class Ticket:
     """A ticket for requesting a Flight stream."""
+
+    cdef:
+        CTicket ticket
+
     def __init__(self, ticket):
-        self.ticket = ticket
+        self.ticket.ticket = tobytes(ticket)
 
     def __repr__(self):
-        return '<Ticket {}>'.format(self.ticket)
+        return '<Ticket {}>'.format(self.ticket.ticket)
+
+
+cdef class Location:
+    """The location of a Flight service."""
+    cdef:
+        CLocation location
+
+    def __init__(self, uri):
+        check_status(CLocation.Parse(tobytes(uri), &self.location))
+
+    def __repr__(self):
+        return '<Location {}>'.format(self.location.ToString())
+
+    @staticmethod
+    def for_grpc_tcp(host, port):
+        """Create a Location for a TCP-based gRPC service."""
+        cdef:
+            c_string c_host = tobytes(host)
+            int c_port = port
+            Location result = Location.__new__(Location)
+        check_status(CLocation.ForGrpcTcp(c_host, c_port, &result.location))
+        return result
+
+    @staticmethod
+    def for_grpc_unix(path):
+        """Create a Location for a domain socket-based gRPC service."""
+        cdef:
+            c_string c_path = tobytes(path)
+            Location result = Location.__new__(Location)
+        check_status(CLocation.ForGrpcUnix(c_path, &result.location))
+        return result
 
+    @staticmethod
+    cdef Location wrap(CLocation location):
+        cdef Location result = Location.__new__(Location)
+        result.location = location
+        return result
 
-class Location(collections.namedtuple('Location', ['host', 'port'])):
-    """A location where a Flight stream is available."""
+    @staticmethod
+    cdef CLocation unwrap(object location):
```

Comment:
You may need to add a `except ?` declaration at the end here, otherwise I don't think exceptions will be propagated to the caller.